File: packages/engine/src/driver/nudge.ts
import { existsSync, readFileSync, writeFileSync } from 'node:fs'

/**
 * The threshold nudge (session-driver 2.3, D4): the one channel a RUNNING
 * headless session can be reached through.
 *
 * Print mode consumes stdin as the prompt and has nothing to listen on
 * afterwards, so the driver speaks through the filesystem instead: the child
 * is launched with `SOFAR_DRIVE_NUDGE` naming a path, `nudge()` creates that
 * file, and the PostToolUse hook — which already runs on every edit and every
 * command — turns its EXISTENCE into "finish the current task, write back,
 * end your turn".
 *
 * Existence is the signal; the contents are detail. A file with unreadable or
 * missing contents still nudges, with the sentence and no number — the fold's
 * tolerance rule applied to a one-line file, and the reason the hook can never
 * fail a session over a half-written nudge.
 *
 * This module is deliberately tiny and dependency-free: it sits on the hook
 * hot path (100ms end to end), where importing the adapter — child_process,
 * readline, the whole spawn surface — to learn one env var name would be
 * absurd. When the env var is absent, which is every session no driver
 * started, `readNudge` costs one property lookup.
 */

/** Env var carrying the nudge file's path into the child. */
export const NUDGE_ENV = 'SOFAR_DRIVE_NUDGE'

/** What the driver saw when it decided to nudge. Every field is optional. */
export interface NudgeDetail {
  /** Percentage of the context window the session was holding. */
  pct?: number
  /** Context tokens behind that percentage. */
  tokens?: number
  ts?: string
}

export function writeNudge(path: string, detail: NudgeDetail = {}): void {
  writeFileSync(path, `${JSON.stringify({ ts: new Date().toISOString(), ...detail })}\n`)
}

/**
 * The nudge this process was given, or null when there is none. Null covers
 * every ordinary session (no env var), a driven session not yet nudged (no
 * file), and an unreadable file — a nudge nobody can read is not one.
 */
export function readNudge(env: NodeJS.ProcessEnv = process.env): NudgeDetail | null {
  const path = env[NUDGE_ENV]
  if (path === undefined || path.length === 0) return null
  try {
    if (!existsSync(path)) return null
    const decoded: unknown = JSON.parse(readFileSync(path, 'utf8'))
    if (typeof decoded !== 'object' || decoded === null || Array.isArray(decoded)) return {}
    const { pct, tokens, ts } = decoded as Record<string, unknown>
    return {
      ...(typeof pct === 'number' && Number.isFinite(pct) ? { pct } : {}),
      ...(typeof tokens === 'number' && Number.isFinite(tokens) ? { tokens } : {}),
      ...(typeof ts === 'string' ? { ts } : {}),
    }
  } catch {
    // Present but unreadable: the signal still stands, the detail does not.
    return {}
  }
}

/**
 * What the hook injects. One line, imperative, and specific about what
 * "finish" means here — a session told only that context is short will
 * summarise and stop, leaving the task half-done and the record silent, which
 * is the stall the threshold policy exists to avoid.
 */
export function nudgeLine(detail: NudgeDetail): string {
  const gauge =
    detail.pct !== undefined
      ? ` — context at ${Math.round(detail.pct)}%${detail.tokens !== undefined ? ` (${detail.tokens} tokens)` : ''}`
      : ''
  return (
    `sofar drive${gauge}: finish the CURRENT task now, then hand off. ` +
    'Mark it done with sofar_update_task, write back with sofar_end_session ' +
    '(summary + the single next action), commit, and end your turn. Do not ' +
    'start another task — the driver launches a fresh session for it.'
  )
}

File: crates/sofar-core/src/nudge.rs
//! The driver's threshold nudge (`driver/nudge.ts`, session-driver 2.3):
//! `SOFAR_DRIVE_NUDGE` names a file whose presence says "wrap up".

use crate::date::js_round;
use crate::json::{self, Json, number_to_string};

pub const NUDGE_ENV: &str = "SOFAR_DRIVE_NUDGE";

#[derive(Debug, Clone, Default, PartialEq)]
pub struct NudgeDetail {
    pub pct: Option<f64>,
    pub tokens: Option<f64>,
    pub ts: Option<String>,
}

/// `readNudge`: None when unset or absent; a present-but-unreadable file is
/// the signal without detail.
#[must_use]
pub fn read_nudge() -> Option<NudgeDetail> {
    let path = std::env::var_os(NUDGE_ENV).filter(|p| !p.is_empty())?;
    let bytes = std::fs::read(&path).ok()?;
    let Ok(Json::Obj(o)) = json::parse(&String::from_utf8_lossy(&bytes)) else {
        return Some(NudgeDetail::default());
    };
    let num = |key: &str| o.get(key).and_then(Json::as_f64).filter(|n| n.is_finite());
    Some(NudgeDetail {
        pct: num("pct"),
        tokens: num("tokens"),
        ts: o.get("ts").and_then(Json::as_str).map(str::to_owned),
    })
}

/// `nudgeLine`.
#[must_use]
pub fn nudge_line(detail: &NudgeDetail) -> String {
    let gauge = detail
        .pct
        .map(|pct| {
            format!(
                " — context at {}%{}",
                number_to_string(js_round(pct)),
                detail
                    .tokens
                    .map(|t| format!(" ({} tokens)", number_to_string(t)))
                    .unwrap_or_default()
            )
        })
        .unwrap_or_default();
    format!(
        "sofar drive{gauge}: finish the CURRENT task now, then hand off. Mark it done with sofar_update_task, write back with sofar_end_session (summary + the single next action), commit, and end your turn. Do not start another task — the driver launches a fresh session for it."
    )
}
